### PR TITLE
BTLx Drill feature

### PR DESCRIPTION
**Pull request by chenkasirer**
image

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ ] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/compas_timber/elements/beam.py [hunk 1]
```diff
@@ -467,7 +484,31 @@ def remove_blank_extension(self, joint_key=None):
...
+        if side_index in (0, 2):  # top + bottom
+            xsize = self.blank_length
+            ysize = self.width
+        elif side_index in (1, 3):  # sides
+            xsize = self.blank_length
+            ysize = self.height
+        elif side_index in (4, 5):  # ends
+            xsize = self.width
+            ysize = self.height
+        return PlanarSurface(xsize, ysize, frame=ref_side, name=ref_side.name)
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/elements/beam.py:508]: No. Point (0, 0) in parametric space is at the corner of the surface. which is the origin point of the corresponding reference side.

**Review by obucklin — COMMENTED** [review]
I left some comments. not sure if they are all correct, just want to clarify a few things

File: src/compas_timber/_fabrication/drilling.py [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+        xy_world = ref_surface.point_at(self.start_x, self.start_y)
+
+        # x and y flipped because we want z pointting down into the beam, that'll be the cylinder long direction
+        cylinder_frame = Frame(xy_world, ref_surface.zaxis, -ref_surface.yaxis)
+        cylinder_frame.rotate(math.radians(self.angle), -ref_surface.zaxis, point=xy_world)
+        cylinder_frame.rotate(math.radians(self.inclination), cylinder_frame.yaxis, point=xy_world)
+
+        depth = self.depth
+        if not self.depth_limited:
+            depth = max(beam.width, beam.height)  # make sure it goes through the beam
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py:332]: I hate trigonometry!

File: src/compas_timber/_fabrication/drilling.py [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        # if it is, the drilling is depth limited otherwise it is not
+        # calculate the x and y coordinates of the start point of the drilling
+        # based on the intersection point and the reference side index
+        # create frame using the intersection point and the reference side axes
+        # calculate the angle and inclination of the drilling using the frame and line
+        # create the drilling process using the calculated parameters
+        ref_side_index, xy_point = cls._calculate_ref_side_index(line, beam)
+        depth_limited = cls._is_depth_limited(line, beam)
+        ref_surface = beam.side_as_surface(ref_side_index)
+        depth = cls._calculate_depth(line, ref_surface) if depth_limited else 0.0
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py:199]: as we discussed, my first priority is usually readability. we can optimized for performance if necessary at a later iteration.

File: src/compas_timber/_fabrication/drilling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (obucklin) [reply src/compas_timber/_fabrication/drilling.py:332]: we could just make it a mile long. 
`if tolerance.units == mm: 
    depth = 25.4*12*5280`

File: examples/model/0004_drilling.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+features = []
+diameter = 10
+for d_line in drill_lines:
+    features.append(Drilling.from_line_and_beam(d_line, diameter, beam))
+
+
+beam.add_features(features)
+
+btlx = BTLx(model)
+PATH = r"C:\Users\ckasirer\Documents\Projects\COMPAS Timber\beam_btlx\drilling.btlx"
```
Comment (jf---) [inline examples/model/0004_drilling.py:48]: why not make the path relative to the file location?

File: src/compas_timber/connections/l_miter.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 from compas.geometry import Vector
 from compas.geometry import cross_vectors
 
-from compas_timber.elements.features import CutFeature
+from compas_timber._fabrication import JackRafterCut
```
Comment (jf---) [inline src/compas_timber/connections/l_miter.py:7]: _fabrication -> fabrication?

Comment (chenkasirer) [reply src/compas_timber/connections/l_miter.py:7]: starting with this PR, there are two parallel fabrication related systems (as @papachap put it nicely, the old one is joint-centric and the new one is feature/process-centric). This is an attempt to let them live side-by-side until all feature types are implemented in the new system. 

Eventually, `_fabrication` will get renamed and replace the existing `fabrication`.

File: src/compas_timber/_fabrication/drilling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gonzalocasas) [reply src/compas_timber/_fabrication/drilling.py:332]: I feel the urge to bring up this classic: https://www.youtube.com/watch?v=JYqfVE-fykk

File: src/compas_timber/_fabrication/drilling.py [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+        cylinder_frame.rotate(math.radians(self.inclination), cylinder_frame.yaxis, point=xy_world)
+
+        depth = self.depth
+        if not self.depth_limited:
+            depth = max(beam.width, beam.height)  # make sure it goes through the beam
+        # move the cylinder by half the depth because frame is the center
+        # add a little notch in the start to make sure it does not drown in the surface
+        # compensate for the notch by adding a little bit to the depth. in any case this is just a vizualisation trick
+        tolerance = self.diameter
+        translation_vector = -cylinder_frame.zaxis * (depth * 0.5 - tolerance)
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py]: reworked this part to be more robust and straight forward
image

File: src/compas_timber/_fabrication/drilling.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py]: now this is hilarious!

File: src/compas_timber/_fabrication/drilling.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py]: reworked this, now creating the cylinder in a more intuitive way, from a line and a diameter

File: src/compas_timber/_fabrication/drilling.py [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        assert self.diameter is not None
+        assert self.angle is not None
+        assert self.inclination is not None
+        assert self.depth is not None
+
+        ref_surface = beam.side_as_surface(self.ref_side_index)
+        xy_world = ref_surface.point_at(self.start_x, self.start_y)
+
+        # x and y flipped because we want z pointting down into the beam, that'll be the cylinder long direction
+        cylinder_frame = Frame(xy_world, ref_surface.zaxis, -ref_surface.yaxis)
```
[the remark this answers is not in the document]
Comment (chenkasirer) [reply src/compas_timber/_fabrication/drilling.py:336]: you can't set the normal (`zaxis`) of a `Frame`, you can only set the `xaxis` and `yaxis` then normal is their cross product based on the right-hand rule.

I am rotating here the `xaxis` of the frame and want to make the drill along it. So after doing the rotations, I indirectly set it to be the normal of the new frame.

**Review by obucklin — APPROVED** [review]

**Pull request merged**